### PR TITLE
Rhoop/docker compose

### PR DESCRIPTION
**Pull request by rhoop**
# Goals of PR

Core changes:

- Dockerization of the app

Notes:

- This does not contain the full Jenkins workflow

### CODE REVIEW

File: docker/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,108 @@
+version: '3.8'
+
+services:
+
+  #------- Secrets
+  generator:
+    image: 792926601177.dkr.ecr.us-east-2.amazonaws.com/specular-platform:specular-latest
+    volumes:
+      - /opt/specular/specular/workspace:/specular/workspace
+      # - sp_workspace:/specular/workspace
```
Comment (chalkandpaste) [inline docker/docker-compose.yml]: Keep or remove?

File: docker/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      - /opt/specular/specular/workspace:/specular/workspace
+    restart: on-failure
+    depends_on:
+      l1-geth:
+        condition: service_healthy
+    healthcheck:
+      test: bash -c "[ -f /tmp/.sp_geth_started.lock ]"
+      retries: 10
+      interval: 5s
+      start_period: 20s
```
Comment (chalkandpaste) [inline docker/docker-compose.yml]: I remember we set this at `40s` on our call yesterday, is `20s` working? No worries if this works.

File: docker/docker-compose.yml (hunk 1 above)
Comment (rhoop) [reply docker/docker-compose.yml]: Nice catch. I'll revert it.

**Review by chalkandpaste — APPROVED** [review]
All LGTM. Just realized, a lot of the script work will get superceded by Maciej's work in a few weeks (just a guess), so I'm sorry to see your nice cleanup and additional features die--but it looks good serves our purposes for now.

File: docker/sp-geth.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+# COPY ../sbin /specular/sbin
+# COPY ../services /specular/services
+
+WORKDIR /specular/workspace
+
+COPY --from=build /specular/ops/ /specular/ops/
+COPY --from=build /specular/services/sidecar/build/bin/sidecar  /usr/local/bin/sidecar
+COPY --from=build /specular/ops/build/bin/genesis  /usr/local/bin/genesis
+COPY --from=build /specular/services/cl_clients/magi/target/debug/magi /usr/local/bin/magi
+COPY --from=build /specular/services/el_clients/go-ethereum/build/bin/geth /usr/local/bin/geth
```
Comment (macwis) [inline docker/sp-geth.Dockerfile:54]: this is great starter! we can optimize later, like sp-geth image practically could contain only sp-geth binary

**Review by macwis — APPROVED** [review]

File: config/local_docker/.genesis.env [hunk 4]
```diff
@@ -5,5 +5,6 @@ ROLLUP_CFG_PATH=./sp_rollup.json
 GENESIS_EXPORTED_HASH_PATH=./genesis_hash.json
 
 L1_NETWORK=localhost
-L1_ENDPOINT=ws://0.0.0.0:8545
```
Comment (LEAFERx) [inline config/local_docker/.genesis.env:8]: why change order?

File: config/local_docker/.paths.env [hunk 5]
```diff
@@ -1,7 +1,8 @@
+BIN_DIR=/usr/local/bin
 CONTRACTS_DIR=/specular/contracts
 CONFIG_DIR=/specular/config
-
-# Define binaries
+OPS_GENESIS_BIN=/usr/local/bin/genesis
```
Comment (LEAFERx) [inline config/local_docker/.paths.env:4]: should this be in other config folders?

File: config/local_docker/.sp_magi.env [hunk 6]
```diff
@@ -1,8 +1,8 @@
 NETWORK=./sp_rollup.json
 JWT_SECRET_PATH=./jwt_secret.txt
-L1_RPC_URL=http://l1-geth:8545
-L2_RPC_URL=http://sp-geth:4011
-L2_ENGINE_URL=ws:/sp-geth:4013
+L1_RPC_URL=http://192.168.160.2:8545
```
Comment (LEAFERx) [inline config/local_docker/.sp_magi.env:3]: why the name is not working?

File: config/local_docker/genesis_config.json [hunk 7]
```diff
@@ -13,17 +13,11 @@
     "l1FeeOverhead": "0x1",
     "l1FeeScalar": "0x1",
     "alloc": {
-        "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266": {
-            "balance": "10000000000000000000"
-        },
-        "0x70997970c51812dc3a010c7d01b50e0d17dc79c8": {
-            "balance": "10000000000000000000"
-        },
         "0x3c44cdddb6a900fa2b585dd299e03d12fa4293bc": {
```
Comment (LEAFERx) [inline config/local_docker/genesis_config.json:16]: should we use `--alloc` flag for genesis generation? will these account change each time when generating secrets? cc @ujvl

**Review by LEAFERx — APPROVED** [review]
Can you format the changes you make? Like removing unnecessary empty lines. Also didn't look into script changes. @ujvl can you take a look, especially the lock file etc.

File: config/local_docker/.sp_magi.env (hunk 6 above)
Comment (rhoop) [reply config/local_docker/.sp_magi.env:3]: it doesn't allow short DNS names

File: config/local_docker/genesis_config.json (hunk 7 above)
Comment (rhoop) [reply config/local_docker/genesis_config.json:16]: 🤷

File: config/local_docker/.paths.env (hunk 5 above)
Comment (rhoop) [reply config/local_docker/.paths.env:4]: It is an inheritance issue that needs to be fixed. I can check again.

File: sbin/generate_secrets.sh [hunk 8]
```diff
@@ -28,17 +37,17 @@ reqdotenv "sp_magi" ".sp_magi.env"
...
 SEQUENCER_ADDRESS=$(generate_wallet $SEQUENCER_PK_FILE)
 echo "Generated account (address=$SEQUENCER_ADDRESS, priv_key_path=$SEQUENCER_PK_FILE)"
-if [ $(realpath "$DISSEMINATOR_PK_PATH") != $(realpath "$SEQUENCER_PK_FILE") ]; then
-  guard_overwrite $DISSEMINATOR_PK_PATH
-  echo "Copying sequencer private key to $DISSEMINATOR_PK_PATH"
-  cp $SEQUENCER_PK_FILE $DISSEMINATOR_PK_PATH
+if [ "$DISSEMINATOR_PK_PATH" != "$SEQUENCER_PK_FILE" ]; then
+  echo "$DISSEMINATOR_PK_PATH" "$SEQUENCER_PK_FILE"
+  guard_overwrite $DISSEMINATOR_PK_PATH $AUTO_ACCEPT
+  cat $SEQUENCER_PK_FILE >$DISSEMINATOR_PK_PATH
```
Comment (ujvl) [inline sbin/generate_secrets.sh:50]: why remove `realpath` (or is this due to an incorrect merge)

**Pull request merged**